### ISSUE TITLE
Metrics not sent on close

### ISSUE THREAD

**Issue by tzolov** [micrometer-metrics/prometheus-rsocket-proxy#35]
After upgrading to latest micrometer 1.5.3 and prometheus-rsocket-proxy 1.0.0 (and Boot 2.3.1) the Timers behavior seems to be broken. We still can receive the LongTaskTimer metrics (probable on part of it) but no the Timer metrics!

It looks like the `PrometheusRSocketClient#client.pushAndClose()` doesn't flush the metrics on exist. If after the pushAndClose() the application waits for ~10 sec before it exit, the Timer metrics get sent!

The [task-demo-metrics-prometheus](https://github.com/tzolov/spring-cloud-dataflow-samples/tree/master/monitoring-samples/task-apps/task-demo-metrics-prometheus) can be used to reproduce and debug the problem (pls ignore the outdated readme). The [Main.java](https://github.com/tzolov/spring-cloud-dataflow-samples/blob/taskPrometeus1.0.0-Boot2.3.2/monitoring-samples/task-apps/task-demo-metrics-prometheus/src/test/java/com/example/task/taskdemometrics/Main.java) is meant to use the plain Micrometer API while the 
[TaskDemoMetricsApplication](https://github.com/tzolov/spring-cloud-dataflow-samples/blob/taskPrometeus1.0.0-Boot2.3.2/monitoring-samples/task-apps/task-demo-metrics-prometheus/src/main/java/com/example/task/taskdemometrics/TaskDemoMetricsApplication.java) is a demo Spring Cloud Task, that internally uses Spring Batch. The Spring Cloud Task uses Timers and LongTaskTimer metrics.

1. use the provided [docker-compose-prometheus.yml](https://github.com/tzolov/spring-cloud-dataflow-samples/blob/taskPrometeus1.0.0-Boot2.3.2/monitoring-samples/task-apps/task-demo-metrics-prometheus/src/test/resources/docker-compose-prometheus.yml) to instantiate  test MySQL Db, SCDF server , prometheus rsocket proxy and prometheus DB. 
   ```
   docker-compose -f ./docker-compose-prometheus.yml up
   ```
   Then use the Prometheus UI (http://localhost:9090) to explore the submitted metrics from the test and form the rsocket proxy. 

2. Simple test 
The [Main.java](https://github.com/tzolov/spring-cloud-dataflow-samples/blob/taskPrometeus1.0.0-Boot2.3.2/monitoring-samples/task-apps/task-demo-metrics-prometheus/src/test/java/com/example/task/taskdemometrics/Main.java) test uses the plain Micrometer API to create a Timer called `tzolov.task` and LongTaskTimer called `tzolov.long` and connects the to the rsocket proxy over the 7001 port. When you run the Main.java the `prometheus_proxy_scrape_active_connections` shows a  connection established between the app and the proxy and the `tzolov.long` (e.g. LongTaskTimer) metrics are received but the `tzolov.task` (e.g. Timer) metrics are never sent. This is visible on the following snapshot: 

<img width="1734" alt="MaintTest" src="https://user-images.githubusercontent.com/1351573/89407673-7976ac80-d71f-11ea-9f29-6611344ad462.png">

But If you uncomment the [Thread.sleep(10000)](https://github.com/tzolov/spring-cloud-dataflow-samples/blob/taskPrometeus1.0.0-Boot2.3.2/monitoring-samples/task-apps/task-demo-metrics-prometheus/src/test/java/com/example/task/taskdemometrics/Main.java#L59) after the `pushAndClose` the `tzolov.task` metrics will eventually reach the server. 

3 Test Spring Cloud Task
Build and start the `TaskDemoMetricsApplication`. the build in `application.properties` will ensure it is connected to the provisioned MySQL and prometheus-proxy. Similarly with the previous example only the `spring_cloud_task_active_` metrics (e.g. LongTaskTimer) are sent with no signs of the `spring_cloud_task_` metrics. 

<img width="1728" alt="SCT" src="https://user-images.githubusercontent.com/1351573/89408261-67e1d480-d720-11ea-8915-7ae7e2b9860e.png">

**Comment by ebezdomnikov** [comment]
any updates? 
I've faced the same issue. Actually I've made some experiments and 0.10.4 version works ok, but from 0.10.5 It has behaviour like you've described.  Additionally, I've noticed what after application dies on version prior 0.10.5 I still been able to get the last (pushAndClose) response from my task application with spring_cloud_task_seconds (I assume the Timer) which is very important as it has the last Timer data with duration application (short-lived application).

**Comment by ebezdomnikov** [comment]
Guys, I'm pretty sure what the problem in puchAndClose (https://github.com/micrometer-metrics/prometheus-rsocket-proxy/compare/v0.10.4...v0.10.5#diff-22863f67dbf81420e7e8e4993769f37a915c23b0ce9373e3988e935fef22f029R150-R157), and the problem is because this method return Mono and what method is using as Bean destroy method. I'm not sure what Spring will execute it, and before application dies it doesn't sent the last metrics to rsocket server.

**Comment by ebezdomnikov** [comment]
Guys, here are also some [information](https://stackoverflow.com/a/58521459/2746065). So the problems are:
1. puchAndClose return Mono which is not executed by Spring during Bean destroy (tested)
2. fireAndForget doesn't send the data immediately to the server, which is crucial  for short-lived application. (also tested)
`
public void pushAndClose() {
    final PublicKey key = latestKey.get();
    if (key != null) {
      try {
        sendingSocket
            .fireAndForget(scrapePayload(key))
            .then(Mono.delay(Duration.ofMillis(100)))
            .then(Mono.fromRunnable(this::close))
            .block();
        return;
      } catch (final Exception ignored) {
        // give up, we tried...
      }
    }
    close();
  }
`

**Comment by shakuzen** [comment]
> 1. puchAndClose return Mono which is not executed by Spring during Bean destroy (tested)

Thank you for checking that for us. That seems to be the assumption here that didn't hold true.

I fixed committed something a little different than the code you shared. It wasn't clear to me what purpose the delay served, so I removed it. We also don't want to block indefinitely on shutdown, in general. I've picked 1 second for now. We could make the time to wait for the fire and forget configurable if that would be useful to users.

Let us know if the problem is fixed in the latest snapshots or the upcoming version 1.4.0.

> I'm also thinking maybe it would be great to replace fireAndForget with requestResponse.

Metrics and other forms of telemetry are typically implemented in such a way as to minimize their impact to the application's functioning. Particularly here where we know the client is being closed probably due to the application shutting down, we don't want to prevent shutdown because there's a problem with the metrics. Imagine we get stuck forever trying to send metrics to a backend that isn't up. Users would not be happy with us that they kept starting more and more applications that never shutdown because of the metrics library.
